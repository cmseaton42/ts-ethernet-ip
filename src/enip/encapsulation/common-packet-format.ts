export enum ItemID {
    NULL = 0x00,
    LIST_IDENTITY = 0x0c,
    CONNECTION_BASED = 0xa1,
    CONNECTED_TRANSPORT_PACKET = 0xb1,
    UCMM = 0xb2,
    LIST_SERVICES = 0x100,
    SOCKET_ADDR_O2T = 0x8000,
    SOCKET_ADDR_T2O = 0x8001,
    SEQUENCED_ADDR_ITEM = 0x8002
}

export interface IDataItems {
    TypeID: ItemID;
    data: Buffer;
}

export abstract class CPF {
    /**
    * Builds a Common Packet Formatted Buffer to be
    * Encapsulated.
    */
    public static build(dataItems: IDataItems[]): Buffer {
        // Write Item Count and Initialize Buffer
        let buf = Buffer.alloc(2);
        buf.writeUInt16LE(dataItems.length, 0);

        for (const item of dataItems) {
            const { TypeID, data } = item;

            // Check valid command id passed
            const isCmd = TypeID in ItemID;
            if (!isCmd) throw new Error("Invalid CPF Type ID!");

            const cpfHeader = Buffer.alloc(4);
            const dataBuf = Buffer.from(data);

            // Write type id to first UINT
            cpfHeader.writeUInt16LE(TypeID, 0);

            // Write data length to second UINT
            cpfHeader.writeUInt16LE(dataBuf.length, 2);

            // Build buffer by concatenating all data items in packet
            buf = dataBuf.length > 0 ? Buffer.concat([buf, cpfHeader, dataBuf]) : Buffer.concat([buf, cpfHeader]);
        }

        return buf;
    }

    /**
    * Parses Incoming Common Packet Formatted Buffer
    * and returns an Array of Objects.
    */
    public static parse(buf: Buffer): IDataItems[] {
        const itemCount = buf.readUInt16LE(0);
        const arr = [];

        let ptr = 2;

        for (let i = 0; i < itemCount; i++) {
            // Get Type ID
            const TypeID: ItemID = buf.readUInt16LE(ptr);
            ptr += 2;

            // Get Data Length
            const length = buf.readUInt16LE(ptr);
            ptr += 2;

            // Get Data from Data Buffer
            const data = Buffer.alloc(length);
            buf.copy(data, 0, ptr, ptr + length);

            // Append Gathered Data Object to Return Array
            arr.push({ TypeID, data });

            ptr += length;
        }

        return arr;
    }
}