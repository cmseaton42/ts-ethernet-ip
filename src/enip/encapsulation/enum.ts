export enum Command {
    NOP = 0x00,
    LIST_SERVICES = 0x04,
    LIST_IDENTITY = 0x63,
    LIST_INTERFACES = 0x64,
    REGISTER_SESSION = 0x65, // Begin Session Command
    UNREGISTER_SESSION = 0x66, // Close Session Command
    SEND_RR_DATA = 0x6f, // Send Unconnected Data Command
    SEND_UNIT_DATA = 0x70, // Send Connnected Data Command
    INDICATE_STATUS = 0x72,
    CANCEL = 0x73
}