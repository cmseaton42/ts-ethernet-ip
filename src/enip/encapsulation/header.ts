import { CPF } from './common-packet-format';

/**
 * Parses Encapulation Status Code to Human Readable Error Message.
 */
export const parseStatus = (status: number): string => {
    switch (status) {
        case 0x00:
            return 'SUCCESS';
        case 0x01:
            return 'FAIL: Sender issued an invalid ecapsulation command.';
        case 0x02:
            return 'FAIL: Insufficient memory resources to handle command.';
        case 0x03:
            return 'FAIL: Poorly formed or incorrect data in encapsulation packet.';
        case 0x64:
            return 'FAIL: Originator used an invalid session handle.';
        case 0x65:
            return 'FAIL: Target received a message of invalid length.';
        case 0x69:
            return 'FAIL: Unsupported encapsulation protocol revision.';
        default:
            return `FAIL: General failure - Error: <${status}> occured.`;
    }
};

export enum Commands {
    NOP = 0x00,
    LIST_SERVICES = 0x04,
    LIST_IDENTITY = 0x63,
    LIST_INTERFACES = 0x64,
    REGISTER_SESSION = 0x65, // Begin Session Command
    UNREGISTER_SESSION = 0x66, // Close Session Command
    SEND_RR_DATA = 0x6f, // Send Unconnected Data Command
    SEND_UNIT_DATA = 0x70, // Send Connnected Data Command
    INDICATE_STATUS = 0x72,
    CANCEL = 0x73,
}

export interface IEncapsulationData {
    commandCode: number;
    command: string;
    length: number;
    session: number;
    statusCode: number;
    status: string;
    options: number;
    data: Buffer | null;
}

export abstract class Header {
    /**
     * Builds an ENIP Encapsolated Packet
     */
    public static build(cmd: Commands, session: number = 0x00, data: Buffer = Buffer.from([])): Buffer {
        const isValidCmd = cmd in Commands;
        if (!isValidCmd) throw new Error('Invalid EIP Encapsulation Command Received!');

        const buf = Buffer.from(data);
        const send = {
            cmd,
            length: buf.length,
            session,
            status: 0x00,
            context: Buffer.alloc(8, 0x00),
            options: 0x00,
            data: buf,
        };

        // Initialize header buffer to appropriate length
        const header = Buffer.alloc(24 + send.length);

        // Build header from encapsulation data
        header.writeUInt16LE(send.cmd, 0);
        header.writeUInt16LE(send.length, 2);
        header.writeUInt32LE(send.session, 4);
        header.writeUInt32LE(send.status, 8);
        send.context.copy(header, 12);
        header.writeUInt32LE(send.options, 20);
        send.data.copy(header, 24);

        return header;
    }

    /**
     * Parses an Encapsulated Packet Received from ENIP Target
     */
    public static parse(buf: Buffer): IEncapsulationData {
        const commandCode = buf.readUInt16LE(0);
        const statusCode = buf.readUInt32LE(8);

        const received: IEncapsulationData = {
            commandCode,
            command: Commands[commandCode],
            length: buf.readUInt16LE(2),
            session: buf.readUInt32LE(4),
            statusCode,
            status: parseStatus(statusCode),
            options: buf.readUInt32LE(20),
            data: null,
        };

        // Get Returned Encapsulated Data
        const dataBuffer = Buffer.alloc(received.length);
        buf.copy(dataBuffer, 0, 24);

        received.data = dataBuffer;

        return received;
    }
}
