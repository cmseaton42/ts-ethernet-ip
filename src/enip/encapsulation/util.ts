import { Header, Commands } from './header';
import { CPF, TypeIDs } from './common-packet-format';

/**
 * Returns a Register Session Request String
 */
export const registerSession = (): Buffer => {
    const cmdBuf = Buffer.alloc(4);
    cmdBuf.writeUInt16LE(0x01, 0); // Protocol Version (Required to be 1)
    cmdBuf.writeUInt16LE(0x00, 2); // Opton Flags (Reserved for Future List)

    // Build Register Session Buffer and return it
    return Header.build(Commands.REGISTER_SESSION, 0x00, cmdBuf);
};

/**
 * Returns an Unregister Session Request String
 *
 * @param {number} session - Encapsulation Session ID
 * @returns {string} unregister seeion strings
 */
export const unregisterSession = (session: number): Buffer => {
    // Build Unregister Session Buffer
    return Header.build(Commands.UNREGISTER_SESSION, session);
};

/**
 * Returns a UCMM Encapsulated Packet String
 *
 * @param {number} [timeout=10] - Timeout (sec)
 */
export const sendRRData = (session: number, data: Buffer, timeout: number = 10) => {
    const timeoutBuf = Buffer.alloc(6);
    timeoutBuf.writeUInt32LE(0x00, 0); // Interface Handle ID (Shall be 0 for CIP)
    timeoutBuf.writeUInt16LE(timeout, 4); // Timeout (sec)

    // Enclose in Common Packet Format
    let buf = CPF.build([{ TypeID: TypeIDs.NULL, data: Buffer.from([]) }, { TypeID: TypeIDs.UCMM, data }]);

    // Join Timeout Data with buffer
    buf = Buffer.concat([timeoutBuf, buf]);

    // Build SendRRData Buffer
    return Header.build(Commands.SEND_RR_DATA, session, buf);
};

/**
 * Returns a Connected Message Datagram (Transport Class 3) String
 *
 * @param {number} ConnectionID - Connection ID from FWD_OPEN
 * @param {number} SequenceNumber - Sequence Number of Datagram
 */
export const sendUnitData = (session: number, data: Buffer, ConnectionID: number, SequnceNumber: number): Buffer => {
    const timeoutBuf = Buffer.alloc(6);
    timeoutBuf.writeUInt32LE(0x00, 0); // Interface Handle ID (Shall be 0 for CIP)
    timeoutBuf.writeUInt16LE(0x00, 4); // Timeout (sec) (Shall be 0 for Connected Messages)

    // Enclose in Common Packet Format
    const seqAddrBuf = Buffer.alloc(4);
    seqAddrBuf.writeUInt32LE(ConnectionID, 0);
    const seqNumberBuf = Buffer.alloc(2);
    seqNumberBuf.writeUInt16LE(SequnceNumber, 0);
    const ndata = Buffer.concat([seqNumberBuf, data]);

    let buf = CPF.build([
        {
            TypeID: TypeIDs.CONNECTION_BASED,
            data: seqAddrBuf,
        },
        {
            TypeID: TypeIDs.CONNECTED_TRANSPORT_PACKET,
            data: ndata,
        },
    ]);

    // Join Timeout Data with
    buf = Buffer.concat([timeoutBuf, buf]);

    // Build SendRRData Buffer
    return Header.build(Commands.SEND_UNIT_DATA, session, buf);
};

/**
 * Returns a Connected Message Datagram (Transport Class 3) Buffer
 *  -> Code Provided by GitHub@Penlane
 */
export const listIdentity = (): Buffer => {
    // Build ListIdentity Buffer
    return Header.build(Commands.LIST_IDENTITY, 0x00);
};

/**
 * Returns a ListServices Buffer
 *  -> Code Provided by GitHub@Penlane
 */
export const listServices = (): Buffer => {
    // Build ListServices Buffer
    return Header.build(Commands.LIST_SERVICES, 0x00);
};
