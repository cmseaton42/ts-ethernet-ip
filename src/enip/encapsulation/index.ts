import { CPF, TypeIDs } from './common-packet-format';
import { Header, Commands } from './header';
import { registerSession, unregisterSession, sendRRData, sendUnitData } from './util';

export const encapsulation = {
    CPF,
    TypeIDs,
    Header,
    Commands,
    registerSession,
    unregisterSession,
    sendRRData,
    sendUnitData,
};
