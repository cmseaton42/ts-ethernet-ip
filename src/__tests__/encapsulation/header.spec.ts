import { Commands, Header, parseStatus } from '../../enip/encapsulation/header';

describe('Encapsulation Header', () => {
    describe('Encapsulation Status Parser', () => {
        it("Returns Proper Human Readable String", () => {
            expect(parseStatus(0)).toEqual("SUCCESS");
            expect(parseStatus(0x01)).toEqual(expect.stringContaining("FAIL"));
            expect(parseStatus(0x02)).toEqual(expect.stringContaining("FAIL"));
            expect(parseStatus(0x03)).toEqual(expect.stringContaining("FAIL"));
            expect(parseStatus(0x04)).toEqual(expect.stringContaining("FAIL"));
            expect(parseStatus(0x64)).toEqual(expect.stringContaining("FAIL"));
            expect(parseStatus(0x65)).toEqual(expect.stringContaining("FAIL"));
            expect(parseStatus(0x69)).toEqual(expect.stringContaining("FAIL"));
            expect(parseStatus(1)).toEqual(expect.stringContaining("FAIL"));
            expect(parseStatus(0x45)).toEqual(expect.stringContaining("FAIL"));
        });
    });

    describe('Header Utility', () => {
        it("Generates encapsulation buffer", () => {
            const test1 = Header.build(Commands.REGISTER_SESSION, 0x00, Buffer.from([0x01, 0x00, 0x00, 0x00]));
            const test2 = Header.build(Commands.UNREGISTER_SESSION);

            expect(test1).toMatchSnapshot();
            expect(test2).toMatchSnapshot();
        });

        it('Throws with bad command when building', () => {
            const test = () => Header.build(0x69, 0x00, Buffer.from([0x01, 0x00, 0x00, 0x00])); // Invalid Command
            expect(test).toThrow();
        })

        it('Parses received encapsulation buffer', () => {
            const raw = Header.build(Commands.SEND_RR_DATA, 98705, Buffer.from([0x01, 0x00, 0x00, 0x00]));
            const test = Header.parse(raw);

            expect(test).toMatchSnapshot();
        })
    })
});