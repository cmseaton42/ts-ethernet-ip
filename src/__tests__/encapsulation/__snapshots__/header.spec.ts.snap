// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Encapsulation Header Header Utility Generates encapsulation buffer 1`] = `
Object {
  "data": Array [
    101,
    0,
    4,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Encapsulation Header Header Utility Generates encapsulation buffer 2`] = `
Object {
  "data": Array [
    102,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Encapsulation Header Header Utility Parses received encapsulation buffer 1`] = `
Object {
  "command": "SEND_RR_DATA",
  "commandCode": 111,
  "data": Object {
    "data": Array [
      1,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  "length": 4,
  "options": 0,
  "session": 98705,
  "status": "SUCCESS",
  "statusCode": 0,
}
`;
